finetune.py

```diff
@@ -15,256 +16,272 @@
-		    x = module(x)
-		    if isinstance(module, torch.nn.modules.conv.Conv2d):
-		    	x.register_hook(self.compute_rank)
-		        self.activations.append(x)
-		        self.activation_to_layer[activation_index] = layer
-		        activation_index += 1
-
-		return self.model.classifier(x.view(x.size(0), -1))
-
-	def compute_rank(self, grad):
-		activation_index = len(self.activations) - self.grad_index - 1
-		activation = self.activations[activation_index]
-		values = \
-			torch.sum((activation * grad), dim = 0).\
-				sum(dim=2).sum(dim=3)[0, :, 0, 0].data
-		
-		# Normalize the rank by the filter dimensions
-		values = \
-			values / (activation.size(0) * activation.size(2) * activation.size(3))
-
-		if activation_index not in self.filter_ranks:
-			self.filter_ranks[activation_index] = \
-				torch.FloatTensor(activation.size(1)).zero_().cuda()
-
-		self.filter_ranks[activation_index] += values
-		self.grad_index += 1
-
-	def lowest_ranking_filters(self, num):
-		data = []
-		for i in sorted(self.filter_ranks.keys()):
-			for j in range(self.filter_ranks[i].size(0)):
-				data.append((self.activation_to_layer[i], j, self.filter_ranks[i][j]))
-
-		return nsmallest(num, data, itemgetter(2))
-
-	def normalize_ranks_per_layer(self):
-		for i in self.filter_ranks:
-			v = torch.abs(self.filter_ranks[i])
-			v = v / np.sqrt(torch.sum(v * v))
-			self.filter_ranks[i] = v.cpu()
-
-	def get_prunning_plan(self, num_filters_to_prune):
-		filters_to_prune = self.lowest_ranking_filters(num_filters_to_prune)
-
-		# After each of the k filters are prunned,
-		# the filter index of the next filters change since the model is smaller.
-		filters_to_prune_per_layer = {}
-		for (l, f, _) in filters_to_prune:
-			if l not in filters_to_prune_per_layer:
-				filters_to_prune_per_layer[l] = []
-			filters_to_prune_per_layer[l].append(f)
-
-		for l in filters_to_prune_per_layer:
-			filters_to_prune_per_layer[l] = sorted(filters_to_prune_per_layer[l])
-			for i in range(len(filters_to_prune_per_layer[l])):
-				filters_to_prune_per_layer[l][i] = filters_to_prune_per_layer[l][i] - i
-
-		filters_to_prune = []
-		for l in filters_to_prune_per_layer:
-			for i in filters_to_prune_per_layer[l]:
-				filters_to_prune.append((l, i))
-
-		return filters_to_prune				
+    def __init__(self, model):
+        self.model = model
+        self.reset()
+
+    def reset(self):
+        # self.activations = []
+        # self.gradients = []
+        # self.grad_index = 0
+        # self.activation_to_layer = {}
+        self.filter_ranks = {}
+
+    def forward(self, x):
+        self.activations = []
+        self.gradients = []
+        self.grad_index = 0
+        self.activation_to_layer = {}
+
+        activation_index = 0
+        for layer, (name, module) in enumerate(self.model.features._modules.items()):
+            x = module(x)
+            if isinstance(module, torch.nn.modules.conv.Conv2d):
+                x.register_hook(self.compute_rank)
```

Comment:
self.compute_rank now returns a function (hook). So when self.compute_rank(activation_index) is called, hook (a partial function with the local variable activation_index) is passed in as the call back function for register_hook.

So when the gradients are updated, hook is called, but doesn't need to calculate the activation_index because it's given when you called (self.compute_ranks(INDEX))